streams/src/main/java/org/apache/kafka/streams/processor/internals/TaskManager.java

```diff
@@ -240,7 +240,14 @@ void suspendTasksAndState()  {
         final AtomicReference<RuntimeException> firstException = new AtomicReference<>(null);
 
         firstException.compareAndSet(null, active.suspend());
+        // close all restoring tasks as well and then reset changelog reader;
+        // for those restoring and still assigned tasks, they will be re-created
+        // in addStreamTasks.
+        firstException.compareAndSet(null, active.closeAllRestoringTasks());
```

Comment:
Why do we close those tasks instead of just suspending them? Could we close them only if they are not re-assigned?

Reply:
The life time of a task is:

created -> [initializeStateStores] -> restoring (writes to the initialized state stores) -> [initializeTopology] -> running -> [closeTopology] -> suspended -> [closeStateManager] -> dead

I.e. the restoring tasks do not have topology initialized at all, whereas `suspend` call is just trying to closeTopology.

Reply:
Ok, but why can't we keep restoring task open than, and hope they get reassigned so we can continue restoring them?

Reply:
Yes we can, the issue is that today we clear all the `store-restorers` hence there's an issue.

We can, of course do some optimizations like `do not close restoring tasks, and also do not clear their corresponding restorers as well`, but this is out of the scope of this PR and I want to address it separately.

Reply:
Ack

Reply:
@guozhangwang Can you create a Jira to track this cleanup?

Reply:
Yup: https://issues.apache.org/jira/browse/KAFKA-7985